### PR TITLE
[CARBONDATA-2585][CARBONDATA-2586][Local Dictionary]Local dictionary support for alter table, preaggregate, varchar datatype, alter set and unset

### PR DESCRIPTION
**Pull request by akashrn5**
**What changes were proposed in this pull request?**
In this PR, 
1. local dictionary support is added for alter table, preaggregate, varChar datatype, alter table set and unset command
2. UTs are added for local dictionary load support
All the validations related to above features are taken care in this PR

**How was this patch tested?**
1. All the tests were executed in 3 node cluster.
2. UTs and SDV test cases are added in the same PR

### CODE REVIEW

**Comment by CarbonDataQA** [comment]
Build Success with Spark 2.1.0, Please check CI http://136.243.101.176:8080/job/ApacheCarbonPRBuilder1/6481/

**Comment by CarbonDataQA** [comment]
Build Failed with Spark 2.2.1, Please check CI http://88.99.58.216:8080/job/ApacheCarbonPRBuilder/5312/

**Comment by CarbonDataQA** [comment]
Build Failed  with Spark 2.1.0, Please check CI http://136.243.101.176:8080/job/ApacheCarbonPRBuilder1/6482/

**Comment by CarbonDataQA** [comment]
Build Success with Spark 2.2.1, Please check CI http://88.99.58.216:8080/job/ApacheCarbonPRBuilder/5313/

**Comment by ravipesala** [comment]
SDV Build Success , Please check CI http://144.76.159.231:8080/job/ApacheSDVTests/5403/

**Comment by ravipesala** [comment]
SDV Build Success , Please check CI http://144.76.159.231:8080/job/ApacheSDVTests/5404/

**Comment by CarbonDataQA** [comment]
Build Success with Spark 2.2.1, Please check CI http://88.99.58.216:8080/job/ApacheCarbonPRBuilder/5344/

**Comment by CarbonDataQA** [comment]
Build Success with Spark 2.1.0, Please check CI http://136.243.101.176:8080/job/ApacheCarbonPRBuilder1/6539/

**Comment by CarbonDataQA** [comment]
Build Success with Spark 2.2.1, Please check CI http://88.99.58.216:8080/job/ApacheCarbonPRBuilder/5368/

**Comment by ravipesala** [comment]
SDV Build Success , Please check CI http://144.76.159.231:8080/job/ApacheSDVTests/5443/

**Comment by CarbonDataQA** [comment]
Build Failed  with Spark 2.1.0, Please check CI http://136.243.101.176:8080/job/ApacheCarbonPRBuilder1/6570/

**Comment by CarbonDataQA** [comment]
Build Failed with Spark 2.2.1, Please check CI http://88.99.58.216:8080/job/ApacheCarbonPRBuilder/5398/

File: core/src/main/java/org/apache/carbondata/core/metadata/schema/table/CarbonTable.java [hunk 1]
```diff
@@ -1110,9 +1110,18 @@ private static void setLocalDictInfo(CarbonTable table, TableInfo tableInfo) {
     String localDictionaryThreshold = tableInfo.getFactTable().getTableProperties()
         .get(CarbonCommonConstants.LOCAL_DICTIONARY_THRESHOLD);
     if (null != isLocalDictionaryEnabled) {
-      table.setLocalDictionaryEnabled(Boolean.parseBoolean(isLocalDictionaryEnabled));
+      // if table is altered then, for same key in table properties, multiple values will be
```
Comment (kumarvishal09) [inline core/src/main/java/org/apache/carbondata/core/metadata/schema/table/CarbonTable.java]: remove this code if unused

File: core/src/main/java/org/apache/carbondata/core/util/CarbonUtil.java [hunk 2]
```diff
@@ -3067,44 +3067,56 @@ public static void setLocalDictColumnsToWrapperSchema(List<ColumnSchema> columns
         }
         // if complex column is defined in local dictionary include column, then get the child
         // columns and set the string datatype child type as local dictionary column
-        if (column.getNumberOfChild() > 0 && null != localDictIncludeColumnsOfMainTable) {
-          listOfDictionaryIncludeColumns = localDictIncludeColumnsOfMainTable.split(",");
+        if (column.getNumberOfChild() > 0 && null != localDictIncludeColumns) {
+          listOfDictionaryIncludeColumns = localDictIncludeColumns.trim().split("\\s*,\\s*");
```
Comment (kumarvishal09) [inline core/src/main/java/org/apache/carbondata/core/util/CarbonUtil.java]: If possible move trim and split outside of loop as it output will be same and it is happening in loop which will impact the performance.

File: core/src/main/java/org/apache/carbondata/core/metadata/schema/table/CarbonTable.java [hunk 3]
```diff
@@ -1110,9 +1110,18 @@ private static void setLocalDictInfo(CarbonTable table, TableInfo tableInfo) {
...
+      // appended, so after splitting, get the first value for enable and threshold, and that will
+      // be latest configurations for the table for further local dictionary generation, because for
+      // alter table add operation, local dictionary enable and threshold configuration is not
+      // allowed
+      String[] isLocalDictEnabled = isLocalDictionaryEnabled.split(",");
```
Comment (ravipesala) [inline core/src/main/java/org/apache/carbondata/core/metadata/schema/table/CarbonTable.java]: Please don't handle the default property here. Better handle in parser like don't add default values if the flow through alter

File: core/src/main/java/org/apache/carbondata/core/util/CarbonUtil.java [hunk 4]
```diff
@@ -3041,24 +3041,30 @@ public static String getBlockId(AbsoluteTableIdentifier identifier, String fileP
...
-        mainTableProperties.get(CarbonCommonConstants.LOCAL_DICTIONARY_EXCLUDE);
+      Map<String, String> mainTableProperties, String isLocalDictEnabledForMainTable) {
     String[] listOfDictionaryIncludeColumns = null;
     String[] listOfDictionaryExcludeColumns = null;
+    String localDictIncludeColumns =
+        mainTableProperties.get(CarbonCommonConstants.LOCAL_DICTIONARY_INCLUDE);
+    String localDictExcludeColumns =
+        mainTableProperties.get(CarbonCommonConstants.LOCAL_DICTIONARY_EXCLUDE);
+    if (null != localDictIncludeColumns) {
+      listOfDictionaryIncludeColumns = localDictIncludeColumns.trim().split("\\s*,\\s*");
```
Comment (ravipesala) [inline core/src/main/java/org/apache/carbondata/core/util/CarbonUtil.java]: what is this split required for?

File: integration/spark-common/src/main/scala/org/apache/carbondata/spark/util/CarbonScalaUtil.scala [hunk 5]
```diff
@@ -615,4 +618,114 @@ object CarbonScalaUtil {
...
+   * @param tableProperties
+   * @param localDictColumns
+   */
+  def validateLocalDictionaryColumns(tableProperties: mutable.Map[String, String],
+      localDictColumns: Seq[String]): Unit = {
+    var dictIncludeColumns: Seq[String] = Seq[String]()
+
+    // check if the duplicate columns are specified in table schema
+    if (localDictColumns.distinct.lengthCompare(localDictColumns.size) != 0) {
+      val duplicateColumns = (dictIncludeColumns ++ localDictColumns)
```
Comment (ravipesala) [inline integration/spark-common/src/main/scala/org/apache/carbondata/spark/util/CarbonScalaUtil.scala]: why adding empty Seq here?

File: integration/spark-common/src/main/scala/org/apache/carbondata/spark/util/CarbonScalaUtil.scala [hunk 6]
```diff
@@ -615,4 +618,114 @@ object CarbonScalaUtil {
...
+
+    // check if the same column is present in both dictionary include and local dictionary columns
+    // configuration
+    if (tableProperties.get(CarbonCommonConstants.DICTIONARY_INCLUDE).isDefined) {
+      dictIncludeColumns =
+        tableProperties(CarbonCommonConstants.DICTIONARY_INCLUDE).split(",").map(_.trim)
+      localDictColumns.foreach { distCol =>
+        if (dictIncludeColumns.exists(x => x.equalsIgnoreCase(distCol.trim))) {
+          val duplicateColumns = (dictIncludeColumns ++ localDictColumns)
+            .diff((dictIncludeColumns ++ localDictColumns).distinct).distinct
```
Comment (ravipesala) [inline integration/spark-common/src/main/scala/org/apache/carbondata/spark/util/CarbonScalaUtil.scala]: I think you require common columns not difference columns here

Comment (akashrn5) [reply integration/spark-common/src/main/scala/org/apache/carbondata/spark/util/CarbonScalaUtil.scala]: basically, the variable name should be changed to common column, i will change that, functionally it is getting the common column

**Comment by CarbonDataQA** [comment]
Build Success with Spark 2.2.1, Please check CI http://88.99.58.216:8080/job/ApacheCarbonPRBuilder/5409/

**Comment by ravipesala** [comment]
SDV Build Fail , Please check CI http://144.76.159.231:8080/job/ApacheSDVTests/5477/

**Comment by CarbonDataQA** [comment]
Build Success with Spark 2.1.0, Please check CI http://136.243.101.176:8080/job/ApacheCarbonPRBuilder1/6590/

**Comment by ravipesala** [comment]
SDV Build Success , Please check CI http://144.76.159.231:8080/job/ApacheSDVTests/5480/

**Comment by ravipesala** [comment]
SDV Build Success , Please check CI http://144.76.159.231:8080/job/ApacheSDVTests/5483/

**Comment by CarbonDataQA** [comment]
Build Success with Spark 2.1.0, Please check CI http://136.243.101.176:8080/job/ApacheCarbonPRBuilder1/6600/

**Comment by CarbonDataQA** [comment]
Build Success with Spark 2.2.1, Please check CI http://88.99.58.216:8080/job/ApacheCarbonPRBuilder/5419/

**Comment by CarbonDataQA** [comment]
Build Success with Spark 2.2.1, Please check CI http://88.99.58.216:8080/job/ApacheCarbonPRBuilder/5429/

**Comment by ravipesala** [comment]
@akashrn5 Please rebase

**Comment by CarbonDataQA** [comment]
Build Failed  with Spark 2.1.0, Please check CI http://136.243.101.176:8080/job/ApacheCarbonPRBuilder1/6621/

**Comment by CarbonDataQA** [comment]
Build Failed  with Spark 2.1.0, Please check CI http://136.243.101.176:8080/job/ApacheCarbonPRBuilder1/6628/

**Comment by CarbonDataQA** [comment]
Build Success with Spark 2.2.1, Please check CI http://88.99.58.216:8080/job/ApacheCarbonPRBuilder/5457/

**Comment by ravipesala** [comment]
SDV Build Success , Please check CI http://144.76.159.231:8080/job/ApacheSDVTests/5504/

**Comment by brijoobopanna** [comment]
retest this please

**Comment by akashrn5** [comment]
retest this please

**Comment by akashrn5** [comment]
retest sdv please

**Comment by CarbonDataQA** [comment]
Build Failed with Spark 2.2.1, Please check CI http://88.99.58.216:8080/job/ApacheCarbonPRBuilder/5461/

**Comment by CarbonDataQA** [comment]
Build Failed  with Spark 2.1.0, Please check CI http://136.243.101.176:8080/job/ApacheCarbonPRBuilder1/6637/

**Comment by CarbonDataQA** [comment]
Build Failed with Spark 2.2.1, Please check CI http://88.99.58.216:8080/job/ApacheCarbonPRBuilder/5465/

**Comment by ravipesala** [comment]
SDV Build Success , Please check CI http://144.76.159.231:8080/job/ApacheSDVTests/5510/

**Comment by ravipesala** [comment]
LGTM

File: core/src/test/java/org/apache/carbondata/core/localdictionary/TestLocalDictionaryGenerator.java [hunk 7]
```diff
@@ -0,0 +1,72 @@
...
+import org.apache.carbondata.core.localdictionary.generator.LocalDictionaryGenerator;
+
+import org.junit.Assert;
+import org.junit.Test;
+
+public class TestLocalDictionaryGenerator {
+
+  @Test
+  public void testColumnLocalDictionaryGeneratorWithValidDataWithinThreshold() {
+    LocalDictionaryGenerator generator = new ColumnLocalDictionaryGenerator(1000);
```
Comment (zzcclp) [inline core/src/test/java/org/apache/carbondata/core/localdictionary/TestLocalDictionaryGenerator.java]: @ravipesala @akashrn5 The constructor ColumnLocalDictionaryGenerator(int) has been modified to constructor ColumnLocalDictionaryGenerator(int, int), please update these issues.

Comment (akashrn5) [reply core/src/test/java/org/apache/carbondata/core/localdictionary/TestLocalDictionaryGenerator.java]: @zzcclp pr is raised #2428

**Pull request closed without merging**